platforms/linux/hax_wrapper.c

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright (c) 2018 Kryptos Logic
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *   1. Redistributions of source code must retain the above copyright notice,
+ *      this list of conditions and the following disclaimer.
+ *
+ *   2. Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in the
+ *      documentation and/or other materials provided with the distribution.
+ *
+ *   3. Neither the name of the copyright holder nor the names of its
+ *      contributors may be used to endorse or promote products derived from
+ *      this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "../../include/hax.h"
+#include "../../core/include/hax_core_interface.h"
+#include "../../core/include/ia32.h"
+
+#include <linux/atomic.h>
+#include <linux/mutex.h>
+#include <linux/slab.h>
+#include <linux/spinlock.h>
+#include <linux/spinlock_types.h>
+
+#include <asm/cmpxchg.h>
+
+int default_hax_log_level = 3;
+int max_cpus;
+hax_cpumap_t cpu_online_map;
+
+int hax_log_level(int level, const char *fmt,  ...)
+{
+    char buffer[256];
```

Comment:
Hmm, this is not ideal. I'm pretty sure this buffer is not big enough for some existing log messages. Can we just pass the `va_list` to `printk`? If not, let's add a TODO comment in case someone else sees it and comes up with an elegant solution.

Reply:
I had tried several approaches before this one, and the necessity of adding the `haxm_*:` prefix caused problems in all of them. However, after checking [this article](https://lwn.net/Articles/310194/), I think I have found the perfect solution via the `%v` placeholder:

> `printk("%shaxm_level: %v, level, fmt, args);`

I will apply this patch to all logging-related functions.

Reply:
I see, that's indeed an elegant solution!

Reply:
Brief update: the approach above didn't work as the patch that documented it didn't get merged. However, I have found a very similar approach while browsing the kernel sources that relies on `va_format` struct and the `%pV` placeholder:

```
    struct va_format vaf;
    va_list args;

    // [...]

    vaf.fmt = fmt;
    vaf.va = &args;
    va_start(args, fmt);
    printk("%shaxm_debug: %pV", KERN_DEBUG, &vaf);
    va_end(args);
```

This, along all previous issues, have been fixed in ac265fd.